docs/docs/api/optimizers/GEPA.md

```diff
@@ -68,6 +68,318 @@ highest_score_achieved_per_task = new_prog.detailed_results.highest_score_achiev
+            
+            updated_components[component_name] = result.improved_instruction
+            
+        return updated_components
+
+# Usage - Setup specialized documentation retriever
+import dspy
+from dspy.retrieve.chromadb_rm import ChromadbRM
+
+# Setup retriever for your specialized documentation
+doc_retriever = ChromadbRM(
+    collection_name="instruction_guidelines",
+    persist_directory="./specialized_docs_db",
+    k=5  # Retrieve top 5 relevant documents
+)
+
+# Configure the RAG-enhanced proposer
+gepa = dspy.GEPA(
+    metric=task_specific_metric,
+    reflection_lm=dspy.LM(model="gpt-5", temperature=1.0, max_tokens=32000, api_key=api_key),
+    instruction_proposer=DocumentationEnhancedProposer(doc_retriever),
+    auto="medium"
+)
+```
+
+#### Integration Patterns
+
+**Using Custom Proposer with External LM:**
+```python
+class ExternalLMProposer(ProposalFn):
+    def __init__(self):
+        # Manage your own LM instance
+        self.external_lm = dspy.LM('gemini/gemini-2.5-pro')
+    
+    def __call__(self, candidate, reflective_dataset, components_to_update):
+        updated_components = {}
+        
+        with dspy.context(lm=self.external_lm):
+            # Your custom logic here using self.external_lm
+            for component_name in components_to_update:
+                # ... implementation
+                pass
+                
+        return updated_components
+
+# Can work without GEPA's reflection_lm!
+gepa = dspy.GEPA(
+    metric=my_metric,
+    reflection_lm=None,  # Optional when using custom proposer
+    instruction_proposer=ExternalLMProposer(),
+    auto="medium"
+)
+```
+
+**Best Practices:**
+- **Use the full power of DSPy**: Leverage DSPy components like `dspy.Module`, `dspy.Signature`, and `dspy.Predict` to create your instruction proposer rather than direct LM calls. Consider `dspy.Refine` for constraint satisfaction, `dspy.ChainOfThought` for complex reasoning tasks, and compose multiple modules for sophisticated instruction improvement workflows
+- **Validate component existence**: Always validate that required components exist in `candidate` and `reflective_dataset`  
+- **Enable holistic feedback analysis**: While dspy.GEPA's `GEPAFeedbackMetric` processes one (gold, prediction) pair at a time, instruction proposers receive all examples for a component in batch, enabling cross-example pattern detection and systematic issue identification.
+- **Mind data serialization**: Serializing everything to strings might not be ideal - handle complex input types (like `dspy.Image`) by maintaining their structure for better LM processing
+- **Create detailed signatures**: Write comprehensive signature docstrings with step-by-step analysis guidance and clear requirements for the instruction improvement task
+- **Test thoroughly**: Test your custom proposer with representative failure cases
+- **Control context size**: Mind the number of examples analyzed to prevent context overflow
```

Comment:
This too is not in control of instruction proposer.

Reply:
👍